### PR TITLE
rgw: allow system users to read SLO parts

### PR DESCRIPTION
**Pull request by cbodley**
multisite data sync relies on fetching the object as the system user

Fixes: http://tracker.ceph.com/issues/19027

### CODE REVIEW

**Comment by yehudasa** [comment]
@cbodley @rzarzynski should we be sync the virtual object or the manifest itself? If the latter, we shouldn't iterate over the parts

**Comment by rzarzynski** [comment]
@yehudasa: good point. If someone needs to go with an SLO, then the virtual object is most likely to large to handle it in the usual way. I think we should sync manifests separately.

I'm not sure whether we offer support for Swift's `multipart-manifest=get` at the moment. RadosGW is compiling... ;-)

**Comment by cbodley** [comment]
in testing the sync, it looks like the GET request to fetch the base object is reading all of its data, and also its slo manifest attribute. after sync, when reading that object on the secondary zone, it's seeing that slo attribute and reading data from its parts instead. so we may want to add a special header to the GET request to avoid fetching data from the parts during sync

**Comment by cbodley** [comment]
@rzarzynski thanks for pointing out `multipart-manifest=get`! the `skip_manifest` flag used by swift to implement that is exactly what i needed for sync requests

i've updated the pr so that `RGWRados::fetch_remote_obj()` adds a `rgwx-sync-manifest` header, which sets `skip_manifest` to avoid fetching object data from the slo parts

still needs testing, but pushed for early feedback

**Comment by cbodley** [comment]
updated after further testing. the `skip_manifest` flag was only applying to `RGW_ATTR_USER_MANIFEST` and not `RGW_ATTR_SLO_MANIFEST`. with that fixed, data sync no longer fetches any object data from the manifest object

to test, i uploaded a big file to the master zone:
```
$ dd bs=1M count=1024 if=/dev/zero of=big.iso
$ swift -A http://localhost:8000/auth/1.0 -U test:tester -K testing upload --use-slo -S 128M CONTAINER big.iso
```
after sync, a stat on the rados object on the secondary zone shows size=0:
```
$ bin/rados -c run/c2/ceph.conf stat -p na-2.rgw.buckets.data a393580b-1e6b-487e-93a1-810e35979d1b.4109.1_big.iso                              
na-2.rgw.buckets.data/a393580b-1e6b-487e-93a1-810e35979d1b.4109.1_big.iso mtime 2017-03-13 11:27:30.000000, size 0
```

**Comment by cbodley** [comment]
passed teuthology tests: http://pulpito.ceph.com/cbodley-2017-03-17_11:36:22-rgw-wip-cbodley-testing---basic-mira/

(will rebase)

**Comment by cbodley** [comment]
@yehudasa can you please review? the initial commit `allow system users to read SLO parts` is not necessary for the multisite sync case, but i still think a system user should be able to read an SLO object locally

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -7001,7 +7001,7 @@ int RGWRados::stat_remote_obj(RGWObjectCtx& obj_ctx,
   int ret = conn->get_obj(user_id, info, src_obj, pmod, unmod_ptr,
                       dest_mtime_weight.zone_short_id, dest_mtime_weight.pg_ver,
                       true /* prepend_meta */, true /* GET */, true /* rgwx-stat */,
-                      &cb, &in_stream_req);
+                      false /* sync manifest */, &cb, &in_stream_req);
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: @cbodley why is it `false` here?

Comment (cbodley) [inline src/rgw/rgw_rados.cc]: hmm, i was thinking that stat requests wouldn't make it that far in `RGWGetObj::execute()`. but it looks like the `if (get_type() == RGW_OP_STAT_OBJ)` check comes after the manifest handling, which probably isn't what we want

i'll change this flag to `true`, and propose a change to `RGW_OP_STAT_OBJ` handling in a separate pr

**Comment by cbodley** [comment]
updated, and proposed a change to `RGW_OP_STAT_OBJ` handling in https://github.com/ceph/ceph/pull/14109

**Comment by yehudasa** [comment]
@cbodley lgtm

**Pull request merged**